Use Duotone presets in block duotone attributes

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Uses Duotone presets (where available) when storing duotone for a given block.

Part of https://github.com/WordPress/gutenberg/issues/46547

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently when setitng Duotone on a given block (e.g. `core/image`) the duotone is stored as an array of colors:

`['#000', '#fff']`

However this isn't useful if you switch Themes. Be storing slugs of Duotone presets we can take advantage of portability of color swatches across Themes.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR checks the selected Duotone swatches against known Duotone preset values and if one is found the slug of the matching Duotone is stored in the block attributes.

If no preset is found then an array of custom colors is stored as per `trunk`.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

getdave committed:
Naive implementation

getdave committed:
Refactor to named functions

scruffian commented:
We need to think about how this will intersect with https://github.com/WordPress/gutenberg/pull/48291, since these presets may not be available anymore...

getdave commented:
> We need to think about how this will intersect with #48291, since these presets may not be available anymore...

That's one of the reasons why you can't do what you are trying in #48291. We need all the presets to be available so users can chose them.

getdave commented:
## Todo

- [x] Fix `Clear` action when removing all Duotones.
- [ ] ~Consider moving `style.color.duotone` to `filter.duotone` (or just `duotone`).~ Follow up.
- [ ] To test this PR we'll need to create a Theme Variation which overides one of the Duotone presets and updates the Hex color values. That way when we select the present and then switch to the Theme Variation we will be able to assert the updated preset's color values are applied.
- [x] Change the front of the site implementation to ensure the Duotone is output.

scruffian commented:
I think it would be better to deal with that in this PR, since this PR deals with the rendering of blocks that use specific presets, so we should make sure that we output the necessary SVGs when the block is rendered.

getdave commented:
> I think it would be better to deal with that in this PR, since this PR deals with the rendering of blocks that use specific presets, so we should make sure that we output the necessary SVGs when the block is rendered.

Can you elaborate on how you think this PR needs to change in order to accommodation what you are doing in https://github.com/WordPress/gutenberg/pull/48291?

getdave committed:
Fix order of hooks

getdave committed:
Utilise existing duotone presets on front of site

getdave committed:
Improve self documenting code

getdave committed:
Refactor away redundant color prop of filter_preset

getdave committed:
Satisfy master yoda…

getdave committed:
More commenting

scruffian commented:
> Can you elaborate on how you think this PR needs to change in order to accommodation what you are doing in https://github.com/WordPress/gutenberg/pull/48291?

Sorry, on reflection I think it makes more sense to keep this PR as is and update the approach in the other PR :)

jeryj commented on lib/block-supports/duotone.php:
```diff
@@ -445,15 +442,44 @@ function gutenberg_render_duotone_support( $block_content, $block ) {
+	// 3. 'unset' (remove filter).
+	$duotone_attr = $block['attrs']['style']['color']['duotone'];
+
+	$is_duotone_colors_array = is_array( $duotone_attr ) && count( $duotone_attr ) === 2;
```
What situation would cause the `count()` to be something other than 2?

jeryj replied:
OK - I see an instance of it using [three colors in the `sherbert.json` for Twenty Twenty Three](https://github.com/WordPress/twentytwentythree/blob/trunk/styles/sherbet.json#L9-L13). I'm not sure yet how it's treated differently though if it's using two or three or more colors.

If it's not handled differently, I'd say remove the `count( $duotone_attr ) === 2` check. If it is handled differently, I'd argue we should rename `$is_duotone_colors_array` to something else since if it has three colors, it's still an array, so the current name would be misleading.

ajlende replied:
Technically the 'duo' in 'duotone' is a misnomer—the SVG filter for the duotone feature works with two _or more_ colors.

However 'duotone' is a much more popular search term for gradient mapping. And the UI for adding more colors was trickier than just doing the two because the `<feComponentTransfer>` `table` method that we're using doesn't allow for color-stop points within that gradient—they're always evenly spaced.

@scruffian has an example in his TT3 variation, https://github.com/WordPress/twentytwentythree/pull/40, that uses four colors for this, sort of, holographic effect.

![four color duotone filter](https://user-images.githubusercontent.com/5129775/220755678-6812660e-6dc1-4964-8e2b-6611cb8bbfe6.jpg)

Anyway, I think the `is_array` check is sufficient here. We can remove the `count` check.

ajlende replied:
> I see an instance of it using [three colors in the sherbert.json for Twenty Twenty Three](https://github.com/WordPress/twentytwentythree/blob/trunk/styles/sherbet.json#L9-L13)

Nice! I didn't realize that made it into TT3.

jeryj replied:
Thanks for the context! So, remove the `count` check here, and then use `$is_duotone_colors_array` in place of the [`$filter_key      = is_array( $duotone_attr ) ? …` check](https://github.com/WordPress/gutenberg/pull/48318#discussion_r1114800819) later should work well. I'll put a little commit in.

getdave replied:
Thanks for the context and fix 👍

jeryj commented on lib/block-supports/duotone.php:
```diff
@@ -445,15 +442,44 @@ function gutenberg_render_duotone_support( $block_content, $block ) {
+		// - an array of colors.
+
+		// Build a unique slug for the filter based on the array of colors.
+		$filter_key      = is_array( $duotone_attr ) ? implode( '-', $duotone_attr ) : $duotone_attr;
```
Could `$is_duotone_colors_array` be used here instead of `is_array( $duotone_attr )` or would the `count( $duotone_attr ) === 2` check earlier get in the way here? Sorry for the naïve questions. Trying to learn about duotone more through these questions.

ajlende replied:
This is not a naïve question—the code isn't very clear about what is going on. 

This `is_array` check was making sure the value isn't `'unset'`. Rather than checking for `'unset'`, I thought it would be more future-proof if other special options needed to be added, but it isn't very clear without a code comment.

getdave replied:
Thanks for the fix.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -311,9 +311,6 @@ function gutenberg_get_duotone_filter_id( $preset ) {
  */
 function gutenberg_get_duotone_filter_property( $preset ) {
-	if ( isset( $preset['colors'] ) && 'unset' === $preset['colors'] ) {
-		return 'none';
```
I wish I would have used `'none'` as the string instead of `'unset'` so that string values could just be treated as CSS like the rest of the block supports. But for maintaining backwards compatibility, I don't think that's really an option anymore.

getdave replied:
Just so you know this return value is preserved but I moved it outside of this function so it doesn't need `colors` in the "preset".

jeryj committed:
Remove count() check for duotone color array

Duotone color arrays can be more than 2 colors, and the code doesn't seem to handle 2 colors differently than 3 or more, so the count() === 2 was unnecessary. Since we've already defined the $is_duotone_colors_array, we can use it later instead of redoing another is_array() check.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -445,15 +442,44 @@ function gutenberg_render_duotone_support( $block_content, $block ) {
-	);
-	$filter_property = gutenberg_get_duotone_filter_property( $filter_preset );
-	$filter_id       = gutenberg_get_duotone_filter_id( $filter_preset );
+	// Possible values for duotone attribute:
+	// 1. array('#000000', '#ffffff').
+	// 2. 'preset-slug'.
+	// 3. 'unset' (remove filter).
+	$duotone_attr = $block['attrs']['style']['color']['duotone'];
```
I do think it's worthwhile in this PR to keep the styles separate from the presets rather than moving it to a follow-up.

Conceptually, I think the `style` attribute was supposed to represent inline styles that were applied to a block. We can't just set `<div class="wp-block-image" style="filter:url('data:image/svg+xml...');">...</div>` because kses strips SVGs and data URL filters don't work in Safari anyway. But conceptually, that's what it means.

Whereas, the top-level attributes (`textColor`, `gradient`, etc.) represented presets that were usually applied with a class on the block wrapper. Again, we can't just apply a class to the wrapper because `filter` is a non-inherited CSS property that we want to apply to a subset of the block.

So I think a top-level `duotone` attribute is more appropriate for saving the presets to.

getdave replied:
I agree it's needed...but that's going to involve:

- deprecations
- additional refactoring to work with the new attribute whilst maintaining the backwards compatibility

...I still think I would be better to make an immediate follow up and avoid bloating this PR.

As soon as I feel I would need to use the word `and` in a PR description to accurately reflect its purpose it's become too large.

Is there a reason it _has_ to be done in _this_ PR?

ajlende replied:
Why do you think it would need deprecations or additional backwards compatibility code?

The way I see it, old blocks with the custom colors saved continue to work like they did before. Nothing here gets touched.

New blocks that select a preset have a new attribute that unsets the `style.color.duotone` attribute and sets the new `duotone` attribute.

I think that it should be this PR so that we don't have to deal with backwards compatibility when switching over to storing custom filters in `style` and presets in `duotone`.

getdave replied:
Eeeek. Too late. But I'll do a follow up 😶

getdave replied:
So we should end up with

- `attributes.filters.duotone` - stores duotone presets only.
- `attributes.style.color.duotone` - stores custom duotone only.

Is that right?

getdave committed:
Improve comments and linting

getdave committed:
Allow Duotone to be cleared using UI

draganescu reviewed (commented):
This is very good and moves things significantly both in terms of the goal of the PR but also in terms of making the code more readable and easier to grok.

getdave commented:
Feel free to review this one. I'm going to add unit tests for the x2 utility functions.

scruffian commented:
I found that if I added a custom Duotone, it didn't work correctly in the frontend. This was the Duotone I added:

```"style":{"color":{"duotone":["rgb(218, 32, 32)","#FFFFCC"]}}```

getdave committed:
Add initial tests for getColorsFromDuotonePreset

getdave committed:
Be explicit about handling presets vs colors

Previously the decision about how to handle values were deferred to the utility. Instead the consuming code should decide whether to look for a preset.

getdave committed:
Complete tests for getColorsFromDuotonePreset

getdave committed:
Add tests and fix implementation of getDuotonePresetFromColors

draganescu commented:
I tested this and it applied my custom duotone somehow wrong:

<details>
<summary>My custom duotone</summary>

<img width="671" alt="Screenshot 2023-02-23 at 11 58 35" src="https://user-images.githubusercontent.com/107534/220875301-06e2900e-248d-4a66-a16d-0d09defb2dd2.png">

</details>


<details>
<summary>Render of the custom duotone</summary>

<img width="673" alt="Screenshot 2023-02-23 at 11 58 45" src="https://user-images.githubusercontent.com/107534/220875389-d501e73f-5a4f-4c59-b383-db1e2b7d0037.png">

</details>

<details>
<summary>Original image</summary>

<img width="669" alt="Screenshot 2023-02-23 at 11 58 58" src="https://user-images.githubusercontent.com/107534/220875519-720a34a1-6615-4b32-84a1-ba9ec8dd6cfc.png">

</details>

getdave committed:
Fix bug with selecting custom colors introduced during test refactor

getdave committed:
Fix custom colors Duotone on front end rendering

getdave commented:
@draganescu @scruffian Thanks for catching custom color breakage. I've reinstated some code I accidentally removed during a refactor. Points to this code being brittle and requiring tests.

draganescu commented:
Re testing the problem described [here](https://github.com/WordPress/gutenberg/pull/48318#issuecomment-1441481141) is gone ✅

scruffian reviewed (approved):
LGTM

youknowriad commented:
So I'm not entirely certain but I believe this PR might have degraded the "typing" performance by 5ms (which is like 10%). 

<img width="578" alt="Screenshot 2023-02-24 at 10 06 49 AM" src="https://user-images.githubusercontent.com/272444/221137865-fdaa96d2-ce1b-448a-acc0-4e528ace11cc.png">

I think that's substantial for a single PR. So I'm wondering whether we should try to revert this in a PR to see the impact it has and potentially find the root cause of the regression.
